NeoML/src/TraditionalML/KMeansClustering.cpp

```diff
@@ -55,9 +55,729 @@ static CPtr<CDnnBlob> createWeightBlob( IMathEngine& mathEngine, const IClusteri
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+// Task which processes a set of elements one-by-one on a single thread
+struct CKMeansClustering::IThreadSubTask : public CKMeansClustering::IThreadTask {
+public:
+	int Size() const override final { return Matrix ? Matrix->Height : VectorSize; }
+
+	const CFloatMatrixDesc* const Matrix;
+protected:
+	IThreadSubTask( const CKMeansClustering& owner, int size ) :
+		IThreadTask( owner, size ),
+		Matrix( nullptr )
+	{}
+	IThreadSubTask( const CKMeansClustering& owner, const CFloatMatrixDesc* matrix ) :
+		IThreadTask( owner, 0 ),
+		Matrix( matrix )
+	{}
+
+	void Run( int threadIndex, int index, int count ) override final;
+	virtual void RunOnElement( int threadIndex, int index ) = 0;
+};
+
+void CKMeansClustering::IThreadSubTask::Run( int threadIndex, int index, int count )
+{
+	const int after_last = index + count;
+	for( int i = index; i < after_last; ++i ) {
+		RunOnElement( threadIndex, i );
+	}
+}
+
+//-------------------------------------------------------------------------------------------------------------
+
+struct CKMeansClustering::CClassifyAllThreadTask final : public CKMeansClustering::IThreadSubTask {
+	CClassifyAllThreadTask( const CKMeansClustering& owner, const CFloatMatrixDesc& matrix ) :
+		IThreadSubTask( owner, &matrix ),
+		Inertia( ThreadCount() )
+	{ DataCluster.SetBufferSize( Size() ); }
+
+	CArray<int> DataCluster{}; // the cluster for this element
+	CInertia Inertia;
+protected:
+	void RunOnElement( int threadIndex, int index ) override
+	{ DataCluster[index] = Owner.findNearestCluster( *Matrix, index, Inertia.Get( threadIndex ) ); }
+};
+
+//-------------------------------------------------------------------------------------------------------------
+
+// Recalculate cluster centers
+struct CKMeansClustering::CUpdateClustersThreadTask final : public CKMeansClustering::IThreadTask {
+	CUpdateClustersThreadTask( const CKMeansClustering& owner,
+		const CFloatMatrixDesc& matrix, const CArray<double>& weights, const CArray<int>& data );
+
+	int Complexity() const { return DataClusters.Size() * Matrix.Width; }
+	bool Reduction() const;
+	bool RunOneThread();
+
+	CArray<CClusterCenter> OldCenters{};
+protected:
+	void Run( int threadIndex, int index, int count ) override;
+
+	const CFloatMatrixDesc& Matrix;
+	const CArray<double>& Weights;
+	const CArray<int>& DataClusters;
+	CArray<bool> IsChanged{};
+};
+
+CKMeansClustering::CUpdateClustersThreadTask::CUpdateClustersThreadTask( const CKMeansClustering& owner,
+		const CFloatMatrixDesc& matrix, const CArray<double>& weights, const CArray<int>& data ) :
+	IThreadTask( owner, owner.clusters.Size() ),
+	Matrix( matrix ),
+	Weights( weights ),
+	DataClusters( data )
+{
+	IsChanged.SetSize( ThreadCount() );
+	OldCenters.SetSize( Size() );
+}
+
+bool CKMeansClustering::CUpdateClustersThreadTask::RunOneThread()
+{
+	// Store the old cluster centers
```

Comment:
Why not
```
Run( 0, 0, DataClusters.Size() );
return IsChanged[0];
```

Reply:
Were you able do detect this speed decrease? Because it looks a bit too much. 2 integer comparisons while `CCommonCluster::Add` iterates over the whole vector length.

Reply:
Ok, I will fix